src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java

```diff
@@ -1024,4 +1029,49 @@ default void onBlockExploded(BlockState state, World world, BlockPos pos, Explos
         world.setBlockState(pos, Blocks.AIR.getDefaultState(), 3);
         getBlock().onExplosionDestroy(world, pos, explosion);
     }
+
+    /**
+     * Gets the slot that the given state takes up when placed in the same block space as other blocks.<br/>
+     *
+     * Returning {@link BlockSlot#FULL_BLOCK} signifies that the given state does not support multipart behavior and
+     * is the value returned by default.
+     */
+    default IBlockSlot getSlot(BlockState state)
+    {
+        return BlockSlot.FULL_BLOCK;
+    }
+
+    /**
+     * Gets the shape for a given state of this block when performing an occlusion test.<br/>
+     * The state may not yet be in the world, so perform any necessary checks.<br/>
+     *
+     * This is the part of the block that cannot and will not be intersected by other parts in the same block space.<br/>
+     * In the case of a pipe or tube, this should only return the center of said block, since the sides can be covered to
+     * block the connection.<br/>
+     */
+    default VoxelShape getOcclusionShape(BlockState state, IBlockReader world, BlockPos pos)
+    {
+        return getBlock().getShape(state, world, pos, ISelectionContext.dummy());
+    }
+
+    /**
+     * Performs an occlusion test between a state of this block and another state.<br/>
+     * One (either) of them may not be in the world, so perform any necessary checks.<br/>
+     *
+     * Returning {@link ActionResultType#SUCCESS} indicates that the blocks do not intersect with each other.<br/>
+     * Returning {@link ActionResultType#FAIL} indicates that the blocks intersect with each other.<br/>
+     * Returning {@link ActionResultType#PASS} indicates that this block does not perform special occlusion testing, and
+     * the other block should also be tested. If both return {@link ActionResultType#PASS}, a default occlusion test is
+     * to be performed by the caller.
+     *
+     * @param state The current state
+     * @param world The current world
+     * @param pos The current position
+     * @param otherState The other state (may or may not be in the world already)
+     * @return Whether the occlusion test succeeded or failed, or is not handled in any special way
+     */
+    default ActionResultType testOcclusion(BlockState state, IBlockReader world, BlockPos pos, BlockState otherState)
```

Comment:
Can we add a note saying fail has priority over success in the javadoc?

Reply:
Good point!